compiler/rustc_ty_utils/src/needs_drop.rs

```diff
@@ -284,27 +284,35 @@ where
                             && component.drop_method_finalizer_elidable(tcx, self.typing_env);
 
                         if self.analysis_kind.is_finalization() {
+                            if adt_def.did()
+                                == tcx.get_diagnostic_item(sym::non_finalizable).unwrap()
+                            {
+                                continue;
+                            }
+                            if adt_def.did() == tcx.get_diagnostic_item(sym::HashMap).unwrap()
```

Comment:
Dumb question: if someone defined their own `struct HashMap` would this match?

Reply:
No. This uses internal rustc magic called 'diagnostic items'. The tl;dr is they work a bit like `lang_item`s but are usually used internally for special-casing error messages. [This](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#227) is how they're defined.